File: src/ast.rs
use chrono::NaiveDate;
use core::fmt;

use crate::parse::parse_atom;

#[derive(Debug, Clone)]
pub enum Atom {
    String(String),
    Number(i32),
    Float(f64),
    Boolean(bool),
    Variable(String),
    Date(NaiveDate),
    DateTime(String),
    Semver(u32, u32, u32),
    // Timestamp(i64)
}

/// Try to interpret a float as semver components (e.g. 5.4 → (5, 4, 0)).
fn float_to_semver(f: f64) -> Option<(u32, u32, u32)> {
    let s = format!("{}", f);
    if let Some((maj_s, min_s)) = s.split_once('.') {
        let maj = maj_s.parse::<u32>().ok()?;
        let min = min_s.parse::<u32>().ok()?;
        Some((maj, min, 0))
    } else {
        let maj = s.parse::<u32>().ok()?;
        Some((maj, 0, 0))
    }
}

impl PartialEq<Atom> for Atom {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Atom::String(s1), Atom::String(s2)) => s1 == s2,
            (Atom::Variable(v1), Atom::Variable(v2)) => v1 == v2,
            (Atom::String(v1), Atom::Variable(v2)) => v1 == v2,
            (Atom::Variable(v1), Atom::String(v2)) => v1 == v2,
            (Atom::Number(n1), Atom::Number(n2)) => n1 == n2,
            (Atom::Float(f1), Atom::Float(f2)) => f1 == f2,
            (Atom::Boolean(b1), Atom::Boolean(b2)) => b1 == b2,
            (Atom::Date(d1), Atom::Date(d2)) => d1 == d2,
            (Atom::DateTime(t1), Atom::DateTime(t2)) => t1 == t2,
            (Atom::Semver(a1, b1, c1), Atom::Semver(a2, b2, c2)) => {
                a1 == a2 && b1 == b2 && c1 == c2
            }
            (Atom::Semver(a, b, c), Atom::Float(f))
            | (Atom::Float(f), Atom::Semver(a, b, c)) => {
                if let Some((maj, min, patch)) = float_to_semver(*f) {
                    *a == maj && *b == min && *c == patch
                } else {
                    false
                }
            }
            (Atom::Semver(a, b, c), Atom::Number(n))
            | (Atom::Number(n), Atom::Semver(a, b, c)) => {
                if *n >= 0 {
                    *a == *n as u32 && *b == 0 && *c == 0
                } else {
                    false
                }
            }
            _ => false,
        }
    }
}

impl PartialOrd for Atom {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        match self {
            Atom::Number(v) => match other {
                Atom::Number(v2) => Some(v.cmp(v2)),
                Atom::Float(v2) => f64::from(*v).partial_cmp(v2),
                Atom::Semver(a2, b2, c2) => {
                    if *v < 0 { return None; }
                    let maj = *v as u32;
                    Some(maj.cmp(a2).then(0u32.cmp(b2)).then(0u32.cmp(c2)))
                }
                _ => None,
            },
            Atom::Float(v) => match other {
                Atom::Float(v2) => v.partial_cmp(v2),
                Atom::Number(v2) => v.partial_cmp(&f64::from(*v2)),
                Atom::Semver(a2, b2, c2) => {
                    let (maj, min, patch) = float_to_semver(*v)?;
                    Some(maj.cmp(a2).then(min.cmp(b2)).then(patch.cmp(c2)))
                }
                _ => None,
            },
            Atom::Date(v) => match other {
                Atom::Date(v2) => v.partial_cmp(v2),
                // TODO: if compare to number it might be unix-timestamp
                _ => None,
            },
            Atom::Semver(a1, b1, c1) => match other {
                Atom::Semver(a2, b2, c2) => {
                    Some(a1.cmp(a2).then(b1.cmp(b2)).then(c1.cmp(c2)))
                }
                Atom::Float(f) => {
                    let (maj, min, patch) = float_to_semver(*f)?;
                    Some(a1.cmp(&maj).then(b1.cmp(&min)).then(c1.cmp(&patch)))
                }
                Atom::Number(n) => {
                    if *n < 0 { return None; }
                    let maj = *n as u32;
                    Some(a1.cmp(&maj).then(b1.cmp(&0).then(c1.cmp(&0))))
                }
                _ => None,
            },
            _ => None,
        }
    }
}

impl fmt::Display for Atom {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Atom::String(string) => write!(f, "{string}"),
            Atom::Number(number) => write!(f, "{number}"),
            Atom::Float(float) => write!(f, "{float}"),
            Atom::Boolean(bool) => write!(f, "{bool}"),
            Atom::Variable(var) => write!(f, "{var}"),
            Atom::Date(var) => write!(f, "{var}"),
            Atom::DateTime(var) => write!(f, "{var}"),
            Atom::Semver(major, minor, patch) => write!(f, "{major}.{minor}.{patch}"),
        }
    }
}

impl<'a> From<&'a str> for Atom {
    fn from(val: &'a str) -> Self {
        let res = parse_atom(val);
        if let Ok((_i, out)) = res {
            return out;
        }
        Atom::String(val.into())
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ComparisonOp {
    Eq,
    More,
    Less,
    MoreEq,
    LessEq,
    NotEq,
    // RegexMatch =~
    // NotRegexMatch !=~
    // Contains =*
    // DoesNotContain !=*
}

impl ComparisonOp {
    pub fn build_from_str(expr: &str) -> Self {
        match expr {
            "==" | "=" => ComparisonOp::Eq,
            ">" => ComparisonOp::More,
            ">=" => ComparisonOp::MoreEq,
            "<" => ComparisonOp::Less,
            "<=" => ComparisonOp::LessEq,
            "!=" | "<>" => ComparisonOp::NotEq,
            _ => unreachable!(),
        }
    }
}
impl fmt::Display for ComparisonOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ComparisonOp::Eq => write!(f, "=="),
            ComparisonOp::More => write!(f, ">"),
            ComparisonOp::Less => write!(f, "<"),
            ComparisonOp::MoreEq => write!(f, ">="),
            ComparisonOp::LessEq => write!(f, "<="),
            ComparisonOp::NotEq => write!(f, "<>"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum LogicOp {
    And,
    Or,
}

impl LogicOp {
    pub fn build_from_str(i: &str) -> Self {
        match i.to_lowercase().as_str() {
            "and" | "&&" => LogicOp::And,
            "or" | "||" => LogicOp::Or,
            _ => unreachable!(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ArrayOp {
    In,
    NotIn,
}

#[derive(Debug, Clone, PartialEq)]
pub enum FnCall {
    Upper,
    Lower,
    Now,
}

#[derive(Debug, Clone, PartialEq)]
pub enum AstNode {
    Void,
    Variable(Atom),
    Function(FnCall, Box<AstNode>),
    Constant(Atom),
    List(Vec<Atom>),
    Compare(Box<AstNode>, ComparisonOp, Box<AstNode>),
    Array(Box<AstNode>, ArrayOp, Box<AstNode>),
    Logic(Box<AstNode>, LogicOp, Box<AstNode>),
    Scope { expr: Box<AstNode>, negate: bool },
}

impl AstNode {
    pub fn as_str(&self) -> Option<&str> {
        match self {
            AstNode::Variable(Atom::Variable(s)) => Some(s.as_str()),
            AstNode::Constant(Atom::String(s)) => Some(s.as_str()),
            AstNode::Constant(Atom::Variable(s)) => Some(s.as_str()),
            _ => None,
        }
    }
}

File: flagfile-ts/src/ast.ts
// ── Atom: leaf values in expressions ────────────────────────────────

export type Atom =
    | { type: 'String'; value: string }
    | { type: 'Number'; value: number }      // integer (i32)
    | { type: 'Float'; value: number }        // f64
    | { type: 'Boolean'; value: boolean }
    | { type: 'Variable'; value: string }
    | { type: 'Date'; value: string }         // YYYY-MM-DD
    | { type: 'DateTime'; value: string }
    | { type: 'Semver'; major: number; minor: number; patch: number };

// ── Operators ──────────────────────────────────────────────────────

export enum ComparisonOp {
    Eq = 'Eq',
    NotEq = 'NotEq',
    More = 'More',
    MoreEq = 'MoreEq',
    Less = 'Less',
    LessEq = 'LessEq',
}

export enum LogicOp {
    And = 'And',
    Or = 'Or',
}

export enum ArrayOp {
    In = 'In',
    NotIn = 'NotIn',
}

export enum FnCall {
    Upper = 'Upper',
    Lower = 'Lower',
    Now = 'Now',
}

// ── AST nodes ──────────────────────────────────────────────────────

export type AstNode =
    | { type: 'Void' }
    | { type: 'Variable'; atom: Atom }
    | { type: 'Function'; fn: FnCall; arg: AstNode }
    | { type: 'Constant'; atom: Atom }
    | { type: 'List'; items: Atom[] }
    | { type: 'Compare'; left: AstNode; op: ComparisonOp; right: AstNode }
    | { type: 'Array'; left: AstNode; op: ArrayOp; right: AstNode }
    | { type: 'Logic'; left: AstNode; op: LogicOp; right: AstNode }
    | { type: 'Scope'; expr: AstNode; negate: boolean };

// ── Flag return values ─────────────────────────────────────────────

export type FlagReturn =
    | { type: 'OnOff'; value: boolean }
    | { type: 'Json'; value: unknown }
    | { type: 'Integer'; value: number }
    | { type: 'Str'; value: string };

// ── Rules ──────────────────────────────────────────────────────────

export type Rule =
    | { type: 'Value'; value: FlagReturn }
    | { type: 'BoolExpressionValue'; expr: AstNode; value: FlagReturn };

// ── Parsed flag map ────────────────────────────────────────────────

export type FlagValue = Map<string, Rule[]>;

// ── Helper constructors ────────────────────────────────────────────

export function atomString(v: string): Atom {
    return { type: 'String', value: v };
}
export function atomNumber(v: number): Atom {
    return { type: 'Number', value: v };
}
export function atomFloat(v: number): Atom {
    return { type: 'Float', value: v };
}
export function atomBoolean(v: boolean): Atom {
    return { type: 'Boolean', value: v };
}
export function atomVariable(v: string): Atom {
    return { type: 'Variable', value: v };
}
export function atomDate(v: string): Atom {
    return { type: 'Date', value: v };
}
export function atomDateTime(v: string): Atom {
    return { type: 'DateTime', value: v };
}
export function atomSemver(major: number, minor: number, patch: number): Atom {
    return { type: 'Semver', major, minor, patch };
}

// ── Atom comparison helpers (mirrors Rust PartialEq / PartialOrd) ──

function floatToSemver(f: number): [number, number, number] | null {
    const s = String(f);
    const dot = s.indexOf('.');
    if (dot === -1) {
        const maj = Math.floor(f);
        if (maj < 0 || !Number.isInteger(f)) return null;
        return [maj, 0, 0];
    }
    const majStr = s.slice(0, dot);
    const minStr = s.slice(dot + 1);
    const maj = parseInt(majStr, 10);
    const min = parseInt(minStr, 10);
    if (isNaN(maj) || isNaN(min)) return null;
    return [maj, min, 0];
}

function atomToNumber(a: Atom): number | null {
    if (a.type === 'Number') return a.value;
    if (a.type === 'Float') return a.value;
    return null;
}

export function atomEquals(a: Atom, b: Atom): boolean {
    // Same type comparisons
    if (a.type === 'String' && b.type === 'String') return a.value === b.value;
    if (a.type === 'Variable' && b.type === 'Variable') return a.value === b.value;
    if (a.type === 'String' && b.type === 'Variable') return a.value === b.value;
    if (a.type === 'Variable' && b.type === 'String') return a.value === b.value;
    if (a.type === 'Number' && b.type === 'Number') return a.value === b.value;
    if (a.type === 'Float' && b.type === 'Float') return a.value === b.value;
    if (a.type === 'Boolean' && b.type === 'Boolean') return a.value === b.value;
    if (a.type === 'Date' && b.type === 'Date') return a.value === b.value;
    if (a.type === 'DateTime' && b.type === 'DateTime') return a.value === b.value;

    // Semver comparisons
    if (a.type === 'Semver' && b.type === 'Semver') {
        return a.major === b.major && a.minor === b.minor && a.patch === b.patch;
    }

    // Semver ↔ Float
    if (a.type === 'Semver' && b.type === 'Float') {
        const sv = floatToSemver(b.value);
        return sv !== null && a.major === sv[0] && a.minor === sv[1] && a.patch === sv[2];
    }
    if (a.type === 'Float' && b.type === 'Semver') {
        const sv = floatToSemver(a.value);
        return sv !== null && b.major === sv[0] && b.minor === sv[1] && b.patch === sv[2];
    }

    // Semver ↔ Number
    if (a.type === 'Semver' && b.type === 'Number') {
        return b.value >= 0 && a.major === b.value && a.minor === 0 && a.patch === 0;
    }
    if (a.type === 'Number' && b.type === 'Semver') {
        return a.value >= 0 && b.major === a.value && b.minor === 0 && b.patch === 0;
    }

    return false;
}

/**
 * Returns negative if a < b, 0 if equal, positive if a > b, or null if
 * the two atoms are not comparable.
 */
export function atomCompare(a: Atom, b: Atom): number | null {
    // Number ↔ Number
    if (a.type === 'Number' && b.type === 'Number') return a.value - b.value;
    // Number ↔ Float
    if (a.type === 'Number' && b.type === 'Float') return a.value - b.value;
    if (a.type === 'Float' && b.type === 'Number') return a.value - b.value;
    // Float ↔ Float
    if (a.type === 'Float' && b.type === 'Float') return a.value - b.value;

    // Date ↔ Date (lexicographic works for YYYY-MM-DD)
    if (a.type === 'Date' && b.type === 'Date') {
        if (a.value < b.value) return -1;
        if (a.value > b.value) return 1;
        return 0;
    }

    // Semver ↔ Semver
    if (a.type === 'Semver' && b.type === 'Semver') {
        const cmp = (a.major - b.major) || (a.minor - b.minor) || (a.patch - b.patch);
        return cmp;
    }

    // Semver ↔ Float
    if (a.type === 'Semver' && b.type === 'Float') {
        const sv = floatToSemver(b.value);
        if (sv === null) return null;
        return (a.major - sv[0]) || (a.minor - sv[1]) || (a.patch - sv[2]);
    }
    if (a.type === 'Float' && b.type === 'Semver') {
        const sv = floatToSemver(a.value);
        if (sv === null) return null;
        return (sv[0] - b.major) || (sv[1] - b.minor) || (sv[2] - b.patch);
    }

    // Semver ↔ Number
    if (a.type === 'Semver' && b.type === 'Number') {
        if (b.value < 0) return null;
        return (a.major - b.value) || (a.minor - 0) || (a.patch - 0);
    }
    if (a.type === 'Number' && b.type === 'Semver') {
        if (a.value < 0) return null;
        return (a.value - b.major) || (0 - b.minor) || (0 - b.patch);
    }

    return null;
}

export function atomToString(a: Atom): string {
    switch (a.type) {
        case 'String': return a.value;
        case 'Number': return String(a.value);
        case 'Float': return String(a.value);
        case 'Boolean': return String(a.value);
        case 'Variable': return a.value;
        case 'Date': return a.value;
        case 'DateTime': return a.value;
        case 'Semver': return `${a.major}.${a.minor}.${a.patch}`;
    }
}
